admin/class-primary-term-admin.php

```diff
@@ -0,0 +1,195 @@
+		$primary_category = $primary_category->get_primary_term();
+
+		if ( false !== $primary_category && $primary_category !== $category->cat_ID ) {
+			$category = get_category( $primary_category );
+		}
+
+		return $category;
+	}
+
+	/**
+	 * Returns an array suitable for use in the javascript
+	 *
+	 * @param stdClass $taxonomy
+	 * @return array
+	 */
+	private function map_taxonomies_for_js( $taxonomy ) {
+		$terms = get_terms( $taxonomy->name );
+
+		$primary_term = new WPSEO_Primary_Term( $taxonomy->name, get_the_ID() );
+
+		return array(
+			'title'   => $taxonomy->labels->singular_name,
+			'name'    => $taxonomy->name,
+			'primary' => $primary_term->get_primary_term(),
+			'terms'   => array_map( array( $this, 'map_terms_for_js' ), $terms ),
+		);
+	}
+
+	/**
+	 * Returns an array suitable for use in the javascript
+	 *
+	 * @param stdClass $term
+	 *
+	 * @return array
+	 */
+	private function map_terms_for_js( $term ) {
+		return array(
+			'id'   => $term->term_id,
+			'name' => $term->name,
+		);
+	}
+
+	/**
+	 * Save the primary term for a specific taxonomy
+	 *
+	 * @param int      $post_ID
+	 * @param stdClass $taxonomy
+	 */
+	private function save_primary_term( $post_ID, $taxonomy ) {
+		$primary_term = filter_input( INPUT_POST, WPSEO_Meta::$form_prefix . 'primary_' . $taxonomy->name . '_term', FILTER_SANITIZE_NUMBER_INT );
+
+		if ( null === $primary_term ) {
+			return;
+		}
+
+		if ( ! check_admin_referer( 'save-primary-term', WPSEO_Meta::$form_prefix . 'primary_' . $taxonomy->name . '_nonce' ) ) {
+			return;
+		}
+
+		$primary_term_object = new WPSEO_Primary_Term( $taxonomy->name, $post_ID );
+		$primary_term_object->set_primary_term( $primary_term );
+	}
+
+	/**
+	 * Returns all the taxonomies for which the primary term selection is enabled
+	 *
+	 * @param int $post_ID Default current post ID.
+	 * @return array
+	 */
+	private function get_primary_term_taxonomies( $post_ID = null ) {
+
+		if ( false !== ( $taxonomies = wp_cache_get( 'primary_term_taxonomies_' . $post_ID, 'wpseo' ) ) ) {
+			return $taxonomies;
+		}
+
+		$post_type      = get_post_type( $post_ID );
+		$all_taxonomies = get_object_taxonomies( $post_type, 'objects' );
+		$all_taxonomies = array_filter( $all_taxonomies, array( $this, 'filter_hierarchical_taxonomies' ) );
+		$taxonomies     = array_filter( $all_taxonomies, array( $this, 'filter_category_taxonomy' ) );
+
+		/**
+		 * Filters which taxonomies for which the user can choose the primary term. Only category is enabled by default.
+		 *
+		 * @api array $taxonomies An array of taxonomy objects that are primary_term enabled.
+		 * @param string $post_type The post type for which to filter the taxonomies.
+		 * @param array $all_taxonomies All taxonomies for this post types, even ones that don't have primary term
+		 *                              enabled.
+		 */
+		$taxonomies = apply_filters( 'wpseo_primary_term_taxonomies', $taxonomies, $post_type, $all_taxonomies );
```

Comment:
Any benefit in casting the result of this `apply_filters()` to an array, so you definitely know that what's cached is an array?